text/0000-const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,750 @@
+- Feature Name: const_generic_const_fn_bounds
+- Start Date: 2018-10-05
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow `impl const Trait` for trait impls where all method impls are checked as const fn.
+
+Make it legal to declare trait bounds on generic parameters of const functions and allow
+the body of the const fn to call methods on the generic parameters that have a `const` modifier
+on their bound.
+
+# Motivation
+[motivation]: #motivation
+
+Without this RFC one can declare const fns with generic parameters, but one cannot add trait bounds to these
+generic parameters. Thus one is not able to call methods on the generic parameters (or on objects of the
+generic parameter type), because they are fully unconstrained.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+You can mark trait implementations as having only `const fn` methods.
+Instead of adding a `const` modifier to all methods of a trait impl, the modifier is added to the trait
+of the `impl` block:
+
+```rust
+struct MyInt(i8);
+impl const Add for MyInt {
+    fn add(self, other: Self) -> Self {
+        MyInt(self.0 + other.0)
+    }
+}
+```
+
+You cannot implement both `const Add` and `Add` for any type, since the `const Add`
+impl is used as a regular impl outside of const contexts. Inside a const context, you can now call
+this method, even via its corresponding operator:
+
+```rust
+const FOO: MyInt = MyInt(42).add(MyInt(33));
+const BAR: MyInt = MyInt(42) + MyInt(33);
+```
+
+You can also call methods of generic parameters of a const function, because they are implicitly assumed to be
+`const fn`. For example, the `Add` trait bound can be used to call `Add::add` or `+` on the arguments
+with that bound.
+
+```rust
+const fn triple_add<T: Add<Output=T>>(a: T, b: T, c: T) -> T {
+    a + b + c
+}
+```
+
+The obligation is passed to the caller of your `triple_add` function to supply a type which has a
+`const Add` impl.
```

Comment:
If I don't call `triple_add` in a const context, does `T` need a `const Add` impl ? Or is it enough for `T` to have an `Add` impl ?

Reply:
An example of what's ok and what isn't would have been helpful here.

Reply:
I am discussing this in depth in https://github.com/rust-lang/rfcs/blob/fcf00a2197908a96391f8a580379214ee22b1b51/text/0000-const-generic-const-fn-bounds.md#runtime-uses-dont-have-const-restrictions

I didn't want to overwhelm the user from the get go.

Reply:
> Is this ok?

yes